Data/Graph/Inductive/Query/SP/DAG.hs

```diff
@@ -0,0 +1,200 @@
+    -> IO Path
+shortestPathBetweenNodes s d l t g = do
+    (_,predecessor) <- runPaths (Just s) (Number 0) Infinity (>) l t g
+    return $ tracePath d predecessor
+
+-- | Returns the longest path between a start node and an end node.
+longestPathBetweenNodes :: (Graph gr, Real b, Foldable f)
+    => Node -- ^ Start
+    -> Node -- ^ Destination
+    -> Int  -- ^ length of topsort
+    -> f Node -- ^ topsort
+    -> gr a b -- ^ Graph
+    -> IO Path
+longestPathBetweenNodes s d l t g = do
+    (_,predecessor) <- runPaths (Just s) (Number 0) NegativeInfinity (<) l t g
+    return $ tracePath d predecessor
+
+-- | Returns the longest possible path from a source node.
+longestPathFromSource :: (Graph gr, Real b, Foldable f)
+    => Node -- ^ Source node
+    -> Int  -- ^ Length of topsort
+    -> f Node -- ^ Topsort
+    -> gr a b -- ^ Graph
+    -> IO Path
+longestPathFromSource s l t g = do
+    (dist,predecessor) <- runPaths (Just s) (Number 0) NegativeInfinity (<) l t g
+    return $ tracePath (maxIndex dist) predecessor
+
+-- | Returns the longest possible path in the graph.
+longestPath :: (Graph gr, Real b, Foldable f)
+    => Int  -- ^ length of topsort
+    -> f Node -- ^ topsort
+    -> gr a b
+    -> IO Path
+longestPath l t g = do
+    (dist,predecessor) <- runPaths Nothing (Number 0) (Number 0) (<) l t g
+    return $ tracePath (maxIndex dist) predecessor
+
+-- | Traces a path backwards through a vector of predecessors,
+--   where a vector of predecessors is a vector such that the value 
+--   in the vector at any position n where n is a node in the graph 
+--   is the predecessor to that node.
+tracePath :: MonadPlus m 
+    => Node -- ^ End node of path
+    -> Vector (Maybe Node) -- ^ Vector of predecessors
+    -> m Node -- ^ MonadPlus of Path
+tracePath n v = 
+    case v ! n of
+         Just node 
+              -> mplus (return n) $ tracePath node v
+         Nothing
+              -> mplus (return n) mzero
+
+-- | A reasonably generalized function used for generating 
+--   a vector of distances and predecessors obtained for a given
+--   comparison operation, an optional source node, and so on.
+runPaths :: (Graph gr, Real b, Foldable f)
+    => Maybe Node -- Source node
+    -> Number b -- ^ Source node initialization value
+    -> Number b -- ^ Distances vector init values
+    -> (Number b -> Number b -> Bool) -- ^ Weight multiplie
+    -> Int -- ^ Length of the top sort
+    -> f Node -- ^ Topological sorting of the graph in some foldable format
+    -> gr a b -- ^ A graph.
+-- | A vector of distances and predecessors. The distances represent maximum or
+--   minimum distance to a given node, and the predecessors represent the node that
+--   linked to that node on the maximum or minimum path.
+    -> IO (Vector (Number b), Vector (Maybe Node))
+runPaths s v ds c l t g = 
+    freezeVecs =<< runPaths' s v ds c l t g
+
+freezeVecs ::
+    (IOVector a, IOVector b)
+    -> IO (Vector a, Vector b)
+freezeVecs (v1,v2) = do
```

Comment:
Do you mean in the outputs of runPaths, or do you mean as inputs to freezeVecs?

For runPaths, it seems necessary to me since you need both to calculate the longest/shortest paths, unless you have an idea of how else they could both be output.

For freezeVecs, its purely convenience. Is it reccomended to use the uncurry definition you provided? I tend to worry that if the code is too compressed it would be too hard to understand for anyone interested in working on it, but if you think I ought to replace it, I will.